### PR TITLE
New csk reader

### PR DESCRIPTION
**Pull request by KineticTheory**
### Background

* Currently, draco contains `src/compton/Compton.cc`, which is a wrapper around CSK functionality
* CSK as a library can read csk ASCII files and perform interpolation in temperature
* CSK also has exe to produce csk ASCII files
* Replaces #930

### Purpose of Pull Request

* CSK was built to be correct, not performant
* Reading ASCII files and interpolating are slow and memory-intensive in CSK
* This is the first step toward replacing Compton and the need to link to CSK
* Much like draco contains a reader/interpolator for opacity data, this adds functionality to read/interpolate Compton scattering data

### Description of changes

* Add preliminary reader/interpolator for Compton scattering data in `src/compton2`
* Compton2 class reads binary csk file and interpolates in temperature
* Compton2 is significantly faster than Compton (CSK)
* Added exe (CSKrw) to convert csk ASCII to binary, removing unneeded Legendre moments (for out_lin, out_nonlin, in_nonlin) and pre-computing temperature derivatives
* Added tests for CSKrw and Compton2
* This represents only a preliminary functionality; more will be added in the future
* For now, keep both compton_interfaces and Compton_tools.
* C4 changes
  * Provide a `size_t` version of `rtt_c4::broadcast` that is used by this PR.
* Also fix linter issues wrt `README.md`.

### Status

* Reference:
  * [Pre-Merge Code Review](https://github.com/lanl/Draco/wiki/Style-Guide)
  * [CDash Status](https://rtt.lanl.gov/cdash/index.php?project=Draco&filtercount=1&showfilters=1&field1=buildname&compare1=63&value1=-pr)
* Checks
  * [ ] Travis/Appveyor CI checks pass
    * Travis issues (long queue times)
  * [x] Code coverage does not decrease
  * [x] [Valgrind test passes](https://rtt.lanl.gov/CDash/index.php?project=Draco)
  * [x] [Toss3 checks pass](https://rtt.lanl.gov/CDash/index.php?project=Draco)
  * [x] [Trinitite checks pass](https://rtt.lanl.gov/CDash/index.php?project=Draco)
  * [x] Code reviewed/approved, sufficient DbC checks, testing, documentation
* WIP
  * [x] fix build warnings on darwin and ccscs6.

### CODE REVIEW

**Comment by attom** [comment]
LGTM. What else is needed?

**Comment by KineticTheory** [comment]
> 
> 
> LGTM. What else is needed?

I want to fix some build warnings that showed up on ccscs6 and darwin.  We probably won't wait for travis.  See https://rtt.lanl.gov/redmine/issues/2276.

Are you done adding features?

**Comment by attom** [comment]
Yes, done adding features

File: src/compton_tools/Compton_Native.cc [hunk 1]
```diff
@@ -0,0 +1,625 @@
...
+  num_temperatures_ = pack[p++];
+  num_groups_ = pack[p++];
+  num_leg_moments_ = pack[p++];
+  num_evals_ = pack[p++];
+  num_points_ = pack[p++];
+  data_size = pack[p++];
+  Insist(data_size < (1U << 31U), "Unrepresentable with int needed in the broadcast");
+
+  // Derived sizes
+  const auto tsz = static_cast<int>(num_temperatures_);
```
Comment (attom) [inline src/compton_tools/Compton_Native.cc:157]: Now that we have the new broadcast, I don't think I need this cast to int?

Comment (KineticTheory) [reply src/compton_tools/Compton_Native.cc]: Fixed in 6404a2d.

**Comment by KineticTheory** [comment]
It might be worth adding an overload for `rtt_c4::broadcast(std::vector<T>, int, int)`.

File: src/compton_tools/Compton_Native.cc (hunk 1 above)
Comment (KineticTheory) [reply src/compton_tools/Compton_Native.cc]: I removed this block of `ints` and switched to _iterator-ranged-based_ `broadcast`.  Let me know what you think of this.

**Comment by attom** [comment]
Thanks for taking a look and cleaning some of this up, @KineticTheory

**Pull request merged**